From 2003b828be0e0fe774e119922e81ee0bb0d2ecdd Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Wed, 9 Aug 2017 16:10:14 +0800
Subject: [PATCH 1/2] fallback: fix double free of dp

If the boot option recorded in csv is not in a media device path, the
corresponding full device path will be referred for creating the boot
variable.

However, the current code logic always frees the full device path
(full_device_path) and the media device path (dp) separately. In order
to resolve this issue, always check whether dp equals to full_device_path
before freeing dp.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 fallback.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fallback.c b/fallback.c
index 9b64077..2ef1b86 100644
--- a/fallback.c
+++ b/fallback.c
@@ -482,7 +482,7 @@ add_to_boot_list(CHAR16 *dirname, CHAR16 *filename, CHAR16 *label, CHAR16 *argum
 err:
 	if (full_device_path)
 		FreePool(full_device_path);
-	if (dp)
+	if (dp && dp != full_device_path)
 		FreePool(dp);
 	if (fullpath)
 		FreePool(fullpath);
-- 
2.13.2

